File: src/World/Objects/Stars.cpp
/*
   Copyright 2010-2011 Tyslenko Max, Bogatirev Pavel.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "World/Objects/Stars.hpp"

ImageStarsData::ImageStarsData(CL_GraphicContext gc, uint16_t mActualSize)
{
    int width = ScreenResolutionX;
    int height = ScreenResolutionY;

    x = rand() % width;
    y = rand() % (height - (height * mActualSize / 100));
    //y = rand() % height;

    brightness = static_cast<float>((rand() % 10 + 2))/10.0f;

    uint16_t mNumStars = resourceManager().sectionHandle("Stars").get_child_nodes().get_length();
    uint16_t randStar = rand() % mNumStars;
    imageHandle = resourceManager().getImage("Stars", boost::lexical_cast<std::string>(randStar));

    float scale = static_cast<float>(rand()%13 + 9)/100.0f;
    imageHandle.set_scale(scale, scale);
    imageHandle.set_alpha(brightness);
}

Stars::Stars(uint16_t maxStars)
    : mBloomSize(0.0f), t1(0), t2(0)
    , mNight(false), mDrawStars(false), mBigGalaxyBloom(0.0f)
{
    srand(time(NULL));

    mGC = appManager().getGraphic();
    mMaxObjects = maxStars;

    // TODO: scale nebula
    mBigGalaxy = resourceManager().getImage("Nebulas", "big_galaxy");
    mBigGalaxyAlpha = 0.8f; // Max alpha value
    mBigGalaxy.set_alpha(0.0);
    mGalaxyX = (ScreenResolutionX  - mBigGalaxy.get_width()) / 2;

}

void Stars::init()
{
    boost::shared_ptr<Foreground> temp = levelManager().getObjectByType<Foreground>("Foreground");
    if (temp->getType() != "Empty") mActualSize = temp->getActualSize();

    for (uint16_t i=0; i < mMaxObjects; ++i)
        mImageStars.push_back(ImageStarsData(mGC, mActualSize));
}

void Stars::update(float elapsed)
{
    float globalTime = worldManager().getAbsTime();
    if (globalTime <= 10.0f) {
        t1 = 0.0f;
    } else if (globalTime >= 14.0f) {
        t1 = 1.0f;
        mNight = true;
    } else {
        mDrawStars = true;
        t1 = ((globalTime - 10.0f) / (14.0f - 10.0f));
        mBloomSize = t1;
        mBigGalaxyBloom = t1 * 0.5f;
    }

    if (mNight)
    {
        if (globalTime <= 20.8f) {
            t2 = 0.0f;
        } else if (globalTime >= 23.8f) {
            t2 = 1.0f;
            mNight = false;
            mDrawStars = false;
        } else {
            t2 = ((globalTime - 20.8f) / (23.8f - 20.8f));
            mBloomSize = 1.0f - t2;
            mBigGalaxyBloom = 1.0f - t2;
        }
    }

    if (!mDrawStars) return;

    if (mBigGalaxyBloom <= mBigGalaxyAlpha)
        mBigGalaxy.set_alpha(mBigGalaxyBloom);

    if (mNight && mBigGalaxyBloom < mBigGalaxyAlpha)
        mBigGalaxyBloom += 0.007f;
    else if (!mNight && mBigGalaxyBloom < mBigGalaxyAlpha)
        mBigGalaxyBloom -= 0.007f;


    if (mNight)
        mBloomSize += 0.001f;
    else
        mBloomSize -= 0.001f;
}

void Stars::updateVisual(float newX, float newY)
{
    if (!mDrawStars) return;

    mBigGalaxy.draw(mGC, mGalaxyX, 0);

    for (uint16_t i=0; i < mMaxObjects; i++)
    {
        // Stars "blinking"
        if (rand() % 70 == 0) continue;

        ImageStarsData &data2 = mImageStars[i];

        if (mBloomSize <= data2.brightness)
            data2.imageHandle.set_alpha(mBloomSize);

        data2.imageHandle.draw(mGC, data2.x, data2.y);
    }
}

boost::shared_ptr<Object> Stars::ParseStars(CL_DomElement tag)
{
    float max = 0;
    if(tag.has_attribute("maxStars"))
    {
        std::string maxStr = tag.get_attribute("maxStars").c_str();
        max = boost::lexical_cast<float>(maxStr);
    }

    float area = Pixels2Meters(ScreenResolutionX) * Pixels2Meters(ScreenResolutionY);
    return boost::shared_ptr<Object>(new Stars(max * area));
}

bool Stars::checkCollision(CL_Pointf point) { return false; }

CL_Rectf Stars::getRectangle()
{
    return CL_Rectf(0, 0, ScreenResolutionX, ScreenResolutionY);
}

CL_Pointf Stars::getPosition()
{
    return CL_Pointf(0, 0);
}

void Stars::setPosition(CL_Pointf newPos)
{
}

File: bin/media/objects/Stars.py
import Core.LogManager
import Core.ApplicationManager
import Core.WorldManager
import Core.ResourceManager
import World.Objects.TypesManager
from World.Objects.Object import *
import Core.Utils
import random

class StarData:
    def __init__(self):
        self.x = random.randint(0, Core.Utils.ScreenResolutionX)
        # TODO: Calculate or get actualSize here!
        self.y = random.randint(0, Core.Utils.ScreenResolutionY)
        self.brightness = random.uniform(0.2, 1.0)

        # Choose random picture for current star
        numStars = Core.ResourceManager.SectionNumChildren("Stars")
        star = random.randint(0, numStars-1)
        self.imageHandle = Core.ResourceManager.getInstance().GetImage("Stars", str(star))

        scale = random.uniform(0.09, 0.2)
        self.imageHandle.SetScale(scale, scale)
        self.imageHandle.SetAlpha(self.brightness)

class Stars(Object):
    def __init__(self, maxObjects):
        super(Stars, self).__init__()

        # We store drop objects here
        self.mStarData = []

        # Precalculated value
        self.mMaxObjects = int(maxObjects)
        self.mBloomSize = 0.0
        self.t1 = self.t2 = 0
        self.mNight = self.mDrawStars = False

        self.mGC = Core.ApplicationManager.getInstance().GetGraphic()
        self.mWorldManagerHandle = Core.WorldManager.getInstance()

    def init(self):
        # TODO: Read actual size here...
        for i in xrange(self.mMaxObjects-1):
            self.mStarData.append(StarData())

    def update(self, elapsed):
       globalTime = self.mWorldManagerHandle.GetAbsTime()
       if globalTime <= 10.0:
           self.t1 = 0
       elif globalTime >= 14.0:
           self.t1 = 1.0
           self.mNight = True
       else:
           self.mDrawStars = True
           self.t1 = ((globalTime - 10.0) / (14.0 - 10.0))
           self.mBloomSize = self.t1

       if self.mNight:
           if globalTime <= 20.8:
               self.t2 = 0
           elif globalTime >= 23.8:
               self.t2 = 1.0
               self.mNight = False
               self.mDrawStars = False
           else:
               self.t2 = ((globalTime - 20.8) / (23.8 - 20.8))
               self.mBloomSize = 1.0 - self.t2

       # If it's day now skip rendering
       if not self.mDrawStars:
           return

       if self.mNight:
           self.mBloomSize += 0.001
       else:
           self.mBloomSize -= 0.001

    def updateVisual(self, newX, newY):
        if not self.mDrawStars:
            return

        for i in xrange(self.mMaxObjects-1):
            # Blink star
            if random.randint(0, 70) == 0:
                continue

            current = self.mStarData[i]
            if self.mBloomSize <= current.brightness:
                current.imageHandle.SetAlpha(self.mBloomSize)

            current.imageHandle.Draw(self.mGC, current.x, current.y)

    def getPosition(self):
        return CL_Pointf(0, 0)

    def setPosition(self):
        print 'Setting position'

    def getRectangle(self):
        return CL_Rectf(0, 0, Core.Utils.ScreenResolutionX, Core.Utils.ScreenResolutionY)

    def checkCollision(self):
        return False

    @staticmethod
    def ParseStars(node):
        maxStars = 0

        if node.HasAttribute(CL_DomString("maxStars")):
            maxStars = float(node.GetAttribute(CL_DomString("maxStars")).CStr())

        area = Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionX) * \
              Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionY)

        return Stars(area * maxStars)

World.Objects.TypesManager.RegisterParser("StarsObject", Stars.ParseStars)
